File: src_llm_test/mcp_tools.py
import inspect
from typing import get_type_hints

REGISTRY = {}


def mcp_tool(
    name: str | None = None,
    description: str | None = None,
    provides: list[str] | None = None,
    consumes: list[str] | None = None,
    parameters: dict[str, str] | None = None,
):

    def wrapper(fn):
        tool_name = name or fn.__name__

        REGISTRY[tool_name] = {
            "fn": fn,
            "schema": build_schema(fn, tool_name, description),
            "provides": provides or [],
            "consumes": consumes or [],
        }
        return fn
    return wrapper


def build_schema(
    fn,
    name: str,
    description: str | None = None,
    param_desc: dict[str, str] | None = None,
):
    import inspect
    from typing import get_type_hints

    sig = inspect.signature(fn)
    hints = get_type_hints(fn)

    properties: dict[str, dict] = {}
    required: list[str] = []

    for param in sig.parameters.values():
        param_name = param.name
        param_type = hints.get(param_name)

        schema_prop = {
            "type": python_type_to_json(param_type)
        }

        if param_desc and param_name in param_desc:
            schema_prop["description"] = param_desc[param_name]

        properties[param_name] = schema_prop

        if param.default is inspect._empty:
            required.append(param_name)

    return {
        "type": "function",
        "function": {
            "name": name,
            "description": description or (fn.__doc__ or ""),
            "parameters": {
                "type": "object",
                "properties": properties,
                "required": required,
            },
        },
    }




def python_type_to_json(tp):
    return {
        int: "number",
        float: "number",
        str: "string",
        bool: "boolean",
    }.get(tp, "string")

File: src_llm_test_ruby/service/mcp_tools.rb
module LLMTestRuby
  module McpTools
    extend self

    @registry = {}

    def registry
      @registry
    end

    def mcp_tool(name: nil, description: nil, provides: [], consumes: [], parameters: {})
      proc do |fn|
        tool_name = name || fn.name
        @registry[tool_name.to_sym] = {
          fn: fn,
          schema: build_schema(fn, tool_name, description, parameters),
          provides: provides,
          consumes: consumes
        }
      end
    end

    private

    def build_schema(fn, name, description, param_desc)
      properties = {}
      required = []

      # Получаем параметры метода
      fn.parameters.each do |type, param_name|
        param_info = {
          'type' => 'string' # по умолчанию
        }

        if param_desc && param_desc[param_name.to_s]
          param_info['description'] = param_desc[param_name.to_s]
        end

        properties[param_name.to_s] = param_info

        # Если параметр обязателен (не имеет значения по умолчанию)
        required << param_name.to_s if type == :req
      end

      {
        'type' => 'function',
        'function' => {
          'name' => name,
          'description' => description || '',
          'parameters' => {
            'type' => 'object',
            'properties' => properties,
            'required' => required
          }
        }
      }
    end
  end
end
